File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        ArrayList<List<Integer>> list = new ArrayList<>();
        Arrays.sort(nums);
        for(int i = 0; i < nums.length - 2; i++) {
            int a = nums[i] * (-1);
            int l = i + 1, r = nums.length - 1;
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            while (r > l) {
                int sum = nums[l] + nums[r];
                if (sum == a) {
                    ArrayList<Integer> inner = new ArrayList<>();
                    inner.add(nums[i]);
                    inner.add(nums[l]);
                    inner.add(nums[r]);
                    list.add(inner);
                    l++;
                    r--;
                    while (l < r && nums[l] == nums[l - 1]) l++;
                } 
                else if (sum > a) r--;
                else l++;
            }
        }
        return list;
    }
}

File: 0015-3sum/0015-3sum.js
var threeSum = function (nums) {
    const n = nums.length;
    const list = [];
    nums.sort((a, b) => a - b);
    for (let i = 0; i < n - 2; i++) {
        if (i > 0 && nums[i] == nums[i - 1]) continue;
        const target = nums[i] * (-1);
        let l = i + 1, r = n - 1;
        while (l < r) {
            const sum = nums[l] + nums[r];
            if (sum == target) {
                const inner = [];
                inner.push(nums[i]);
                inner.push(nums[l]);
                inner.push(nums[r]);
                list.push(inner);
                l++;
                r--;
                while (nums[l] === nums[l - 1]) l++;
            }
            else if (sum < target) l++;
            else r--;
        }
    }
    return list;
};